File: subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        size = len(nums)
        res = []
        
        def backtrack(idx,temp):
            if idx>=size:
                res.append(temp)
                return 
            
            backtrack(idx+1,temp)
            backtrack(idx+1,temp+[nums[idx]])
            return
        
        backtrack(0,[])
        return res

File: subsets/subsets.java
class Solution {
    
    public static void backtrack(List<List<Integer>> res, int[] nums, int idx, ArrayList<Integer> currList){
        if(idx >= nums.length){
            res.add((ArrayList<Integer>)currList.clone());
            return;
        }
        
        backtrack(res,nums,idx+1,currList);
        currList.add(nums[idx]);
        backtrack(res,nums,idx+1,currList);
        currList.remove(currList.size() - 1);
        return;
    }
    
    public List<List<Integer>> subsets(int[] nums) {
        //List<List<Integer>> res = ArrayList<List<>>();
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        backtrack(res,nums,0,new ArrayList<Integer>());
        return res;
    }
}